src/iter.rs

```diff
@@ -555,95 +551,195 @@ impl<'a> ExactSizeIterator for Chars<'a> {}
     },
+    Light,
 }
 
 impl<'a> Lines<'a> {
     pub(crate) fn new(node: &Arc<Node>) -> Lines {
-        Lines {
-            iter: LinesEnum::Full {
-                node: node,
-                start_char: 0,
-                end_char: node.text_info().chars as usize,
-                start_line: 0,
-                total_line_breaks: node.line_break_count(),
-                line_idx: 0,
-            },
-            is_reversed: false,
-        }
+        let info = node.text_info();
+        Lines::new_with_range(
+            node,
+            (0, info.bytes as usize),
+            (0, info.line_breaks as usize + 1),
+        )
     }
 
     pub(crate) fn new_with_range(
         node: &Arc<Node>,
-        char_idx_range: (usize, usize),
-        line_break_idx_range: (usize, usize),
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
     ) -> Lines {
-        Lines::new_with_range_at(
-            node,
-            line_break_idx_range.0,
-            char_idx_range,
-            line_break_idx_range,
-        )
+        Lines::new_with_range_at_impl::<false>(node, 0, byte_idx_range, line_idx_range)
     }
 
     pub(crate) fn new_with_range_at(
         node: &Arc<Node>,
         at_line: usize,
-        char_idx_range: (usize, usize),
-        line_break_idx_range: (usize, usize),
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
     ) -> Lines {
-        Lines {
-            iter: LinesEnum::Full {
-                node: node,
-                start_char: char_idx_range.0,
-                end_char: char_idx_range.1,
-                start_line: line_break_idx_range.0,
-                total_line_breaks: line_break_idx_range.1 - line_break_idx_range.0 - 1,
-                line_idx: at_line,
-            },
-            is_reversed: false,
+        Lines::new_with_range_at_impl::<true>(node, at_line, byte_idx_range, line_idx_range)
+    }
+
+    fn new_with_range_at_impl<const AT_LINE: bool>(
+        node: &Arc<Node>,
+        line: usize,
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
+    ) -> Lines {
+        let start_byte = byte_idx_range.0;
+        let end_byte = byte_idx_range.1;
+        debug_assert!(node.is_char_boundary(start_byte));
+        debug_assert!(node.is_char_boundary(end_byte));
+
+        // curretly cut off the first line by handeling the same as a full slice
+        if AT_LINE {
+            if line == line_idx_range.0 {
+                return Lines::new_with_range(node, byte_idx_range, line_idx_range);
+            } else if line >= line_idx_range.1 {
+                let mut res = Lines::new_with_range_at(
+                    node,
+                    line_idx_range.1 - 1,
+                    byte_idx_range,
+                    line_idx_range,
+                );
+                res.next();
+                return res;
+            }
+        }
+        // For convenience/readability.
```

Comment:
I believe in the `Chunks` iterator this comment accompanies the declaration of the `start_byte` and `end_byte` variables.